storeliveness: add more logging

miraradeva opened the pull request:
This commit adds additional store liveness logging. At the default logging level, the new logs include all state transitions (e.g. a store received support for a new epoch, a store withdrew support, a store was marked as idle). At vmodule level 2, the logs include some aggregated stats (e.g. number of heartbeats/responses sent), and at vmodule level 3, the logs include all support extensions.

Fixes: [#131530](https://github.com/cockroachdb/cockroach/issues/131530)

Release note: None

miraradeva committed:
storeliveness: add more logging

This commit adds additional store liveness logging. At the default
logging level, the new logs include all state transitions (e.g. a store
received support for a new epoch, a store withdrew support, a store was
marked as idle). At vmodule level 2, the logs include some aggregated
stats (e.g. number of heartbeats/resonses sent), and at vmodule level
3, the logs include all support extensions.

Fixes: #131530

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131673)
<!-- Reviewable:end -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131673)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @miraradeva, @nvanbenschoten, and @xinhaoz)

___
*[`-- commits` line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-O8CUivx5U9ohvQYk90t):*
This issue doesn't contain any motivation for the channel, is it possible to get an explanation here? Are these logs high-frequency and would crowd info out in other channels or fill up files? Is there a reason someone would route this separately?
___
*[`-- commits` line 17 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-O8CTx_50UbMaAqp5-JW):*
Logging channels become part of our public documentation and incur an incremental complexity cost on the customer's end. Can you explain in the release note and channel description what's expected to be output to this channel and how a customer is expected to consume it.

The answer could very well be: ignore this channel. I just want to make that clear for the user.


<!-- Sent from Reviewable.io -->

kevin-v-ngo commented:
I'm also curious on what the use case is for this new logging channel. I'm assuming this is for internal L3 debugging. 

- Is this for SH or Cloud?
- Since the primary touchpoint for log collection is Debug zip for SH, do we expect customers to configure this channel? I'm assuming customer wouldn't want to have this on by default so it'd be a retroactive workflow (turn on and wait for the issue to reoccur) with customers during escalations correct?

miraradeva commented:
Sorry for the confusion, I realized this PR is a bit sparse on justification.

The idea is that Store Liveness will be the new failure detection mechanism powering leader leases; it will, in effect, replace node liveness for leasing purposes. As part of this work, we've been focusing on logging more information about state changes (stores going down, stores becoming partitioned or unreachable for other reasons). Store liveness is a fairly low-level component that will affect the behavior of higher layers, like Raft and leasing, so it will be useful to have a full view of the store liveness state transitions.

The part I'm not completely sure of yet is how much we expect the customer to interact with these new logs. I'll bring this up in our project meeting today.

miraradeva reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131673)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @nvanbenschoten, and @xinhaoz)

___
*[`-- commits` line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-O8CUivx5U9ohvQYk90t):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This issue doesn't contain any motivation for the channel, is it possible to get an explanation here? Are these logs high-frequency and would crowd info out in other channels or fill up files? Is there a reason someone would route this separately?
</blockquote></details>

I added more context in the issue and the release note. Happy to port some of that to the main commit message.
___
*[`-- commits` line 17 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-O8CTx_50UbMaAqp5-JW):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Logging channels become part of our public documentation and incur an incremental complexity cost on the customer's end. Can you explain in the release note and channel description what's expected to be output to this channel and how a customer is expected to consume it.

The answer could very well be: ignore this channel. I just want to make that clear for the user.
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

miraradeva commented:
Ok I added more context and justification for the change in the GH issue, release note, and in the channel description. Happy to expand these even more. Also, after discussing with the team, I renamed the new channel to `LIVENESS` to capture the concept more generally and eventually include node liveness logs too.

To answer some of @dhartunian's and @kevin-v-ngo's questions:

> Are these logs high-frequency and would crowd info out in other channels or fill up files?

In a healthy cluster, these wouldn't be a lot of logs. But in cases of disk stalls and network partitions, they could get verbose. This is also true for the existing node liveness logs.

> Is there a reason someone would route this separately?

The main reason to keep them separate is to be able to get a clear picture of the cluster health from a liveness perspective. The motivation is similar to why we prefer having all distribution logs in kv distribution.

> and incur an incremental complexity cost on the customer's end.

@dhartunian can you expand on this some more? I think we may be missing some of the reasons to hold back on adding new channels.

----

> I'm assuming this is for internal L3 debugging.

Yes, mostly, but customers can find insights from these logs too, just like node liveness logs. We will iterate on the store liveness logs and ensure customer get a clear message from them.

> Is this for SH or Cloud?

Both. Leader leases will eventually become the default lease type, and they depend on store liveness as the failure detection mechanism.

> Since the primary touchpoint for log collection is Debug zip for SH, do we expect customers to configure this channel?

I'm not super familiar with how customers usually configure log collection. I was under the impression that any log file will be included in the debuug zip by default. @kevin-v-ngo am I missing something?

> I'm assuming customer wouldn't want to have this on by default.

We would want this logging on by default. There will be more verbose variants configurable using vmodule but the basic logging will be needed by default, just like node liveness heartbeat logs are needed now.

kevin-v-ngo commented:
> > I'm assuming this is for internal L3 debugging.
> 
> Yes, mostly, but customers can find insights from these logs too, just like node liveness logs. We will iterate on the store liveness logs and ensure customer get a clear message from them.

Thanks for explaining! If the primary use case is for internal L3 debugging I think we should reconsider adding a new channel since customers will most likely not configure them unless their is a clear user case for them. As David mentioned, it's incremental complexity to route and export/externalize these logs to their downstream system. 

Have we seen the [KV_DISTRIBUTION](https://www.cockroachlabs.com/docs/v24.2/logging#kv_distribution) be used at all for escalations? 

If we want this information to always be collected to help with L3 troubleshooting, perhaps we should consider a crdb_internal table + debug zip (we'd need to make sure this table doesn't blow up debug zip) or the system.eventlog (assuming this is a relatively low-frequency event) which we dump into debug zip today. 

> 
> > Is this for SH or Cloud?
> 
> Both. Leader leases will eventually become the default lease type, and they depend on store liveness as the failure detection mechanism.

I mainly asked because we can always have this on by default for cloud clusters. So current feedback is really for Self-hosted escalations where the logs are not collected by default.

> I'm not super familiar with how customers usually configure log collection. I was under the impression that any log file will be included in the debuug zip by default. @kevin-v-ngo am I missing something?
It's unfortunately not. The customer needs to enable and configure the specific logging channel.

miraradeva commented:
Decided to reduce the scope of this PR and not add the new logging channel yet, until we're sure exactly what logs will be useful to separate from the main logs. Created https://github.com/cockroachdb/cockroach/issues/133051 to track.

nvb reviewed (commented):
Sorry for letting this review sit. It mostly looks good, just a few comments.

> Reviewed 17 of 17 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-:-O9uFpZLE9_3iepmeCGA:b-w39979)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @miraradeva, and @xinhaoz)

___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 149 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-O9uDROY63484VpXwMpr:-O9uDROY63484VpXwMps:b23lw2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/f33542ec7b3f497a50e9dca7a3ae944ac0f9c06f/pkg/kv/kvserver/storeliveness/requester_state.go#L149)):*
> ```Go
> 		// this is a new store, recentlyQueried will be set to true in addStore.
> 		if rs.recentlyQueried.Swap(active) == idle {
> 			wasIdle = true
> ```

nit: `wasIdle = rs.recentlyQueried.Swap(active) == idle`
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 421 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-O9uE-qf8sZhnq5ek499:-O9uE-qf8sZhnq5ek49A:bmqc5uj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f33542ec7b3f497a50e9dca7a3ae944ac0f9c06f/pkg/kv/kvserver/storeliveness/requester_state.go#L421)):*
> ```Go
> // others are not interesting enough to log?
> func logSupportFromChange(ctx context.Context, ss slpb.SupportState, ssNew slpb.SupportState) {
> 	if ss.Epoch == ssNew.Epoch && !ss.Expiration.IsEmpty() && !ssNew.Expiration.IsEmpty() {
> ```

Related to your TODO above, I feel like this could be structured in a way which both (1) makes the cases more obviously exhaustive, and (2) makes the logic more closely mirror its meaning. Something like:
```go
if ss.Epoch == ssNew.Epoch {
	if ss.Expiration.IsEmpty() {
		log.Infof(ctx, "received support ...")
	} else {
		log.VInfof(ctx, 3, "extended ...")
	}
} else {
	assert(ss.Epoch < ssNew.Epoch)
	if ssNew.Expiration.IsEmpty() {
		log.Infof(ctx, "lost support from ...")
	} else {
		// TODO: anything?
        }
}
```
___
*[`pkg/kv/kvserver/storeliveness/support_manager.go` line 122 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-O9uFH-sDnIt23g1iZwW:-O9uFH-sDnIt23g1iZwX:b-a76iaf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f33542ec7b3f497a50e9dca7a3ae944ac0f9c06f/pkg/kv/kvserver/storeliveness/support_manager.go#L122)):*
> ```Go
> 		// addStore checks if the store exists before adding it.
> 		sm.storesToAdd.addStore(id)
> 		log.Infof(context.TODO(), "store %+v is not heartbeating store %+v yet", sm.storeID, id)
> ```

Do we need this log and the one in `maybeAddStores`? Will the combination be perceived as spammy?
___
*[`pkg/kv/kvserver/storeliveness/support_manager.go` line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-O9uFStb7aDs8hcMt-Dd:-O9uFStb7aDs8hcMt-De:b-2q7psn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f33542ec7b3f497a50e9dca7a3ae944ac0f9c06f/pkg/kv/kvserver/storeliveness/support_manager.go#L254)):*
> ```Go
> 	for _, store := range sta {
> 		if sm.requesterStateHandler.addStore(store) {
> 			log.Infof(ctx, "starting to heartbeat store %+v", store)
> ```

Does this context have the right log tags so that it's clear from the logs which local store we're talking about?


<!-- Sent from Reviewable.io -->

miraradeva reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131673)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @nvanbenschoten, and @xinhaoz)

___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 149 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-O9uDROY63484VpXwMpr:-OAJ12nGB5kVNLMNnjIY:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/f33542ec7b3f497a50e9dca7a3ae944ac0f9c06f/pkg/kv/kvserver/storeliveness/requester_state.go#L149)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: `wasIdle = rs.recentlyQueried.Swap(active) == idle`
</blockquote></details>

Done
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 421 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-O9uE-qf8sZhnq5ek499:-OAJ3csx9pR6WGxeePrQ:b3qez7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/f33542ec7b3f497a50e9dca7a3ae944ac0f9c06f/pkg/kv/kvserver/storeliveness/requester_state.go#L421)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Related to your TODO above, I feel like this could be structured in a way which both (1) makes the cases more obviously exhaustive, and (2) makes the logic more closely mirror its meaning. Something like:
```go
if ss.Epoch == ssNew.Epoch {
	if ss.Expiration.IsEmpty() {
		log.Infof(ctx, "received support ...")
	} else {
		log.VInfof(ctx, 3, "extended ...")
	}
} else {
	assert(ss.Epoch < ssNew.Epoch)
	if ssNew.Expiration.IsEmpty() {
		log.Infof(ctx, "lost support from ...")
	} else {
		// TODO: anything?
        }
}
```
</blockquote></details>

Ok, I like this better; it mirrors the logic in `handleHeartbeatResponse` and uses the exact same assertions. Makes me wonder if it's worth doing it in a separate function (instead of logging directly in `handleHeartbeatResponse`)
___
*[`pkg/kv/kvserver/storeliveness/support_manager.go` line 122 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-O9uFH-sDnIt23g1iZwW:-OAJ1djn0wUPnh0RKkW0:b-dv9vzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f33542ec7b3f497a50e9dca7a3ae944ac0f9c06f/pkg/kv/kvserver/storeliveness/support_manager.go#L122)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need this log and the one in `maybeAddStores`? Will the combination be perceived as spammy?
</blockquote></details>

This one is useful in recording exactly when `SupportFrom` was called for the first time, and the one in `maybeAddStores` would come after some delay from the queue. Maybe we can keep both for now while investigating timing issues? I can make this one to be at level 2.
___
*[`pkg/kv/kvserver/storeliveness/support_manager.go` line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-O9uFStb7aDs8hcMt-Dd:-OAJ2LVDCbihof-aDpRn:b-3vns4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/f33542ec7b3f497a50e9dca7a3ae944ac0f9c06f/pkg/kv/kvserver/storeliveness/support_manager.go#L254)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this context have the right log tags so that it's clear from the logs which local store we're talking about?
</blockquote></details>

Everything coming from the main loop goroutine has enough context to tell at which store the action is happening. The two places where we don't have a context are the logs in `SupportFrom`. So I've included the local store id in those.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131673)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @miraradeva, and @xinhaoz)

___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 395 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-OAJLMGi6Rcdxa1Grn92:-OAJLMGi6Rcdxa1Grn93:bfh5pmn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbc632ba967c6f578208f4e9afe27c1c38af9c/pkg/kv/kvserver/storeliveness/requester_state.go#L395)):*
> ```Go
> 	}
> 	if ss.Epoch == msg.Epoch {
> 		assert(
> ```

Could this have been a delayed message from before we started requesting support for the epoch? Same question down below.
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 434 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-OAJN1RkBKdAzf4VgE5T:-OAJN1RkBKdAzf4VgE5U:b-v98q1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbc632ba967c6f578208f4e9afe27c1c38af9c/pkg/kv/kvserver/storeliveness/requester_state.go#L434)):*
> ```Go
> 		}
> 	} else {
> 		assert(
> ```

Do we need to duplicate the assertions?
___
*[`pkg/kv/kvserver/storeliveness/supporter_state.go` line 265 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-OAJMDZz7WwAFldIqtTG:-OAJMDZz7WwAFldIqtTH:b-nf0a14) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbc632ba967c6f578208f4e9afe27c1c38af9c/pkg/kv/kvserver/storeliveness/supporter_state.go#L265)):*
> ```Go
> // of a support requester upon receiving a heartbeat.
> func handleHeartbeat(ss slpb.SupportState, msg *slpb.Message) slpb.SupportState {
> 	assert(
> ```

nit throughout this PR and in some existing code: these 3-4 line assertions are pretty visually noisy. They're doubling the size of these methods and making them meaningfully harder to read. Also, in most cases, they aren't even needed to stay under the 100 character recommended line wrap. Consider consolidating them to single lines where possible, like:
```go
assert(!msg.Expiration.IsEmpty(), "requested support with zero expiration")
```


<!-- Sent from Reviewable.io -->

miraradeva reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131673)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @nvanbenschoten, and @xinhaoz)

___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 395 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-OAJLMGi6Rcdxa1Grn92:-OAJcjWI-Z_q9uchW2lN:byrxf3g) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbc632ba967c6f578208f4e9afe27c1c38af9c/pkg/kv/kvserver/storeliveness/requester_state.go#L395)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could this have been a delayed message from before we started requesting support for the epoch? Same question down below.
</blockquote></details>

Yes, and I was missing a unit test for it. 
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 434 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-OAJN1RkBKdAzf4VgE5T:-OAJhyWG9UL7z_GzkTCA:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbc632ba967c6f578208f4e9afe27c1c38af9c/pkg/kv/kvserver/storeliveness/requester_state.go#L434)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need to duplicate the assertions?
</blockquote></details>

Done
___
*[`pkg/kv/kvserver/storeliveness/supporter_state.go` line 265 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-OAJMDZz7WwAFldIqtTG:-OAJhPZL9uHyDALdgyN0:b-yil7b0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbc632ba967c6f578208f4e9afe27c1c38af9c/pkg/kv/kvserver/storeliveness/supporter_state.go#L265)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit throughout this PR and in some existing code: these 3-4 line assertions are pretty visually noisy. They're doubling the size of these methods and making them meaningfully harder to read. Also, in most cases, they aren't even needed to stay under the 100 character recommended line wrap. Consider consolidating them to single lines where possible, like:
```go
assert(!msg.Expiration.IsEmpty(), "requested support with zero expiration")
```
</blockquote></details>

Agreed. I trimmed some of the logging lines too.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131673)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @miraradeva, and @xinhaoz)

___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 434 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-OAJN1RkBKdAzf4VgE5T:-OAJp7fc5oGVeeDrvLNE:b-1k6nhy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbc632ba967c6f578208f4e9afe27c1c38af9c/pkg/kv/kvserver/storeliveness/requester_state.go#L434)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Sorry for being unclear. What I meant was: do we need to have assertions in these `log` functions if we already have them in the corresponding message handlers?
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 414 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-OAJpgI1DkqXcteDQA5O:-OAJpgI22Cr07Ah2AaFw:b-907vzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c3dec37dd86e55746a4fffc2c056f6ecc05e735/pkg/kv/kvserver/storeliveness/requester_state.go#L414)):*
> ```Go
> 			log.Infof(ctx, "received support from %v", supportChangeStr(ss, ssNew))
> 		} else {
> 			log.VInfof(ctx, 3, "extended support from %v", supportChangeStr(ss, ssNew))
> ```

This is going to allocate the `suportChangeStr` even if verbose logging is not enabled. You can wrap this in a `log.ExpensiveLoggingEnabled(ctx, 3)` check to avoid that.

Same comment in `supporter_state.go`.
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 439 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-OAJoqMdEmCv7FuA0dXu:-OAJoqMdEmCv7FuA0dXv:bp5upjn) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c3dec37dd86e55746a4fffc2c056f6ecc05e735/pkg/kv/kvserver/storeliveness/requester_state.go#L439)):*
> ```Go
> 
> func supportChangeStr(current slpb.SupportState, previous slpb.SupportState) string {
> 	return fmt.Sprintf("store %v; current = %+v, previous = %+v", current.Target, current, previous)
> ```

Passing through a `string` is going to break log redaction. Can this use `redact.Sprintf` and return a `redact.RedactableString`?

Related to that point, have you checked whether the logging in this package is sufficiently unreacted to be useful for debugging?


<!-- Sent from Reviewable.io -->

miraradeva reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131673)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @nvanbenschoten, and @xinhaoz)

___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 434 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-OAJN1RkBKdAzf4VgE5T:-OAJyGPN57c8v9d5KVRx:bmgi7te) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbc632ba967c6f578208f4e9afe27c1c38af9c/pkg/kv/kvserver/storeliveness/requester_state.go#L434)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Sorry for being unclear. What I meant was: do we need to have assertions in these `log` functions if we already have them in the corresponding message handlers?
</blockquote></details>

Yeah I agree we don't need all the assertions (and I removed most). But this one seemed useful to illustrate that the new epoch is higher, so we've lost support (and we're logging that). In the actual logic we have more fine-grained assertions. 
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 414 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-OAJpgI1DkqXcteDQA5O:-OAJyCZO55ypLhJqoUAp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c3dec37dd86e55746a4fffc2c056f6ecc05e735/pkg/kv/kvserver/storeliveness/requester_state.go#L414)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is going to allocate the `suportChangeStr` even if verbose logging is not enabled. You can wrap this in a `log.ExpensiveLoggingEnabled(ctx, 3)` check to avoid that.

Same comment in `supporter_state.go`.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 439 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-OAJoqMdEmCv7FuA0dXu:-OAJzQXy9-Dl-D1hymR6:bh5fpya) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c3dec37dd86e55746a4fffc2c056f6ecc05e735/pkg/kv/kvserver/storeliveness/requester_state.go#L439)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Passing through a `string` is going to break log redaction. Can this use `redact.Sprintf` and return a `redact.RedactableString`?

Related to that point, have you checked whether the logging in this package is sufficiently unreacted to be useful for debugging?
</blockquote></details>

They look intact. I ran this to check:
```
./cockroach demo --insecure --vmodule=support_manager=3,requester_state=3,supporter_state=3 --nodes=5 --log='file-defaults: { redactable: true, redact: true, dir: /Users/mradeva/Desktop/logs }'
```


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131673)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @miraradeva, and @xinhaoz)

___
*[`pkg/kv/kvserver/storeliveness/supporter_state.go` line 280 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-OANmXfyBNFBNrlxl6Gn:-OANmXfyBNFBNrlxl6Go:bn876s9) ([raw file](https://github.com/cockroachdb/cockroach/blob/193f5cd001ad34896b4a1948408c5d8e7d1f8506/pkg/kv/kvserver/storeliveness/supporter_state.go#L280)):*
> ```Go
> func logSupportForChange(ctx context.Context, ss slpb.SupportState, ssNew slpb.SupportState) {
> 	assert(!ssNew.Expiration.IsEmpty(), "requested support with zero expiration")
> 	if ss.Epoch == ssNew.Epoch && !ss.Expiration.IsEmpty() && log.ExpensiveLogEnabled(ctx, 3) {
> ```

Aren't we now going to hit the "provided support" branch whenever verbose logging is not enabled?


<!-- Sent from Reviewable.io -->

miraradeva reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131673)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @nvanbenschoten, and @xinhaoz)

___
*[`pkg/kv/kvserver/storeliveness/supporter_state.go` line 280 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-OANmXfyBNFBNrlxl6Gn:-OANyRUS2z6bbwUj6MKE:blzkzc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/193f5cd001ad34896b4a1948408c5d8e7d1f8506/pkg/kv/kvserver/storeliveness/supporter_state.go#L280)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Aren't we now going to hit the "provided support" branch whenever verbose logging is not enabled?
</blockquote></details>

You're right. Sorry for the back and forth. I'll fix the merge conflict in a separate push.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131673#-:-OANydUI8EOsBvw2sI0w:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, and @xinhaoz)



<!-- Sent from Reviewable.io -->

miraradeva commented:
The last push rebased a fresh master (conflicts in `BUILD.bazel` and `support_manager`). And removed an unnecessary if check:
<img width="1148" alt="Screenshot 2024-10-29 at 2 55 20 PM" src="https://github.com/user-attachments/assets/f6c0939f-652d-442f-9bf0-4c6b8cc55b01">

miraradeva commented:
bors r=nvanbenschoten
